core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -274,28 +284,32 @@ public ShardFailedClusterStateTaskExecutor(AllocationService allocationService,
 
                     ShardRouting matched = currentState.getRoutingTable().getByAllocationId(task.shardId, task.allocationId);
                     if (matched == null) {
-                        // tasks that correspond to non-existent shards are marked as successful
-                        logger.debug("{} ignoring shard failed task [{}] (shard does not exist anymore)", task.shardId, task);
-                        batchResultBuilder.success(task);
-                    } else {
-                        // remove duplicate actions as allocation service expects a clean list without duplicates
-                        if (seenShardRoutings.contains(matched)) {
-                            logger.trace("{} ignoring shard failed task [{}] (already scheduled to fail {})", task.shardId, task, matched);
+                        Set<String> inSyncAllocationIds = indexMetaData.inSyncAllocationIds(task.shardId.id());
+                        if (task.primaryTerm > 0 && inSyncAllocationIds.contains(task.allocationId)) {
+                            // marking allocation id without shard routing as stale only allowed if primary term is provided
+                            // otherwise we might remove the allocation id of a primary that was failed in an earlier cluster update task
+                            // when no more active shards are available now.
```

Comment:
I think it would help to clarify two things via comments in the code:

1. under what conditions will the `primaryTerm == 0` (basically a local shard failure), and why is it that we do not care about the primaryTerm value in that scenario?

2. how does a `primaryTerm == 0` indicate that the primary was failed in a previous cluster state update and we are aiming to avoid an empty allocate ID set?  Its unclear to me still how `primaryTerm == 0` signals this situation and in any case, it would be very helpful to have it in code comments.